content/gocourse-2019-05-melbourne.slide

```diff
@@ -813,50 +816,103 @@ Embedding types introduces the possiblity for naming conflicts. Go follows some
+		error
+		Timeout() bool
+		Temporary() bool
 	}
+- Implementation of the `Error` interface here - [[https://golang.org/pkg/net/#AddrError][net/AddrError]]
+- The error values can be logged and printed. Example -
 
-By convention, functions often return `error` as the second/last return value.
-The calling code should handle errors by testing whether the error equals nil:
+	g, err := Sqrt(-1); err != nil {
+		log.Println(err)
+	}
 
-	func getCat(strID string) (*Cat, error) {
-		id, err := strconv.Atoi(strID)
-		if err != nil {
-		    return nil, CatError(err)
-		}
-		return catStore.cats[id], nil
+
+* Error Handling
+
+- Error handling: Check for the presence of an error (compare with nil).
+- Do not compare error variables. ([[https://goplay.space/#mgw9QAmeLUP][Example]])
+- Errors are values so it can be programmed. ([[https://goplay.space/#MTqqxwEdq-A][Example]])
+	g, err := Sqrt(-1); err != nil {
+		log.Println(err)
 	}
+- Handle errors gracefully ([[https://goplay.space/#kpFwYYBE0ND][A useful pattern]])
+- To check if an `error` is of certain type use custom error types with type assertion
+- Example: ([[https://goplay.space/#t4s_BrAQTLt][Errors and type assertion]])
+- By convention many standard library functions return `error` as the second return value.
 
+"Errors are values."  - [[https://go-proverbs.github.io/][Rob Pike]]
+
+* Defer
+
+- `defer` (keyword) allows you to postpone execution.
+- When `defer` statement executes, "the parameters to call are evaluated and saved anew"
+- Deferred functions are called when "function returns" or "stack unwinds".
+- Deferred functions are called in reverse order.
+- Used for resource cleanup (close IO resources, release locks etc.).
+- Syntax
+	defer <expression>
+- Example [[https://goplay.space/#wvwUzU1TpNw][How defer works]]
 
-* Errors (2/3)
 
-"Don't panic." - [[https://go-proverbs.github.io/][Rob Pike]]
+* Panic and Recover
 
-Panic should be used when something went unexpectedly wrong.
-For instance, initialise a regular expression with [[https://golang.org/pkg/regexp/#MustCompile][regexp.MustCompile]] if the pattern is a string constant.
-`MustCompile` will `panic` at initialisation.
+`panic` and `recover` behaves similar to `try/catch` in other languages.
 
-	var re = regexp.MustCompile(`^[a-c]+$`)
-	var re2 = regexp.MustCompile(`^[a-c+$`) // panic
+- `panic` causes
+- |__ stack unwinding
+- |__ calls deferred Functions
+- |__ can pass value into `panic(v)`
 
+- `recover`
+- |__ stops the stack unwinding process
+- |__ must be called from within a deferred function (in the same goroutine)
+- |__ returns the value passed by call to panic
 
-Use `panic` rarely and judiciously in your own code!
+	func panic(interface{})
+	func recover() interface{}
 
 
-* Errors (3/3)
+* Don't Panic
 
-- Error [[https://goplay.space/#h4c7gQxzQsN][create]], [[https://goplay.space/#mze1De36XjV][handle]], [[https://goplay.space/#h1kA3qKaO6B][custom error]]
+- Because panic terminates the program/goroutine.
```

Comment:
From https://blog.golang.org/defer-panic-and-recover:

> Panic is a built-in function that stops the ordinary flow of control…

`panic` does not necessarily terminate the program or goroutine, since `defer` can arrest the stack-unwinding process.

Reply:
Yes that's right. I meant to say it could potentially if there is no recover. And is not a favourable scenario in programs. I've changed the sentence to "Panic could terminate the program / goroutine".